File: src/disassembler/disassembler.py
from typing import List

from disassembly_table import disassembly_table
from operator_disassembly import operator_disassembly_table


def disassemble(byte_code: bytes) -> List[str]:
    assembly: List[str] = []

    print(f'Total bytes: {len(byte_code)}')

    print('LINE\tOPERATOR\t\tASSEMBLY\tSIZE, INDEX')

    index = 0
    line_number = 0
    while index < len(byte_code):

        start_index = index
        operator = byte_code[index]
        index += 1
        
        name, args, sizes, sized_indexes = disassembly_table[operator]
        
        # Check if the operator has sized operands
        if sized_indexes:
            # Get the size of the operands
            handled_size = byte_code[index]
            index += 1
            name += str(handled_size)

            # Update the operand sizes
            new_sizes = list(sizes)
            for arg_index in sized_indexes:
                new_sizes[arg_index] = handled_size
            sizes = tuple(new_sizes)

        string = f'{line_number}:   \t{operator}:{hex(operator)}   \t\t{name}'

        # Disassemble the operands
        for arg, size in zip(args, sizes):
            operand_string = operator_disassembly_table[arg](byte_code[index : index + size])
            index += size
            string += f' {operand_string}'

        string += f'     \t({index - start_index}, {start_index})'
        assembly.append(string)

    return assembly
        

File: disassembler/src/disassembler.rs
use rust_vm_lib::assembly::{ByteCode, AssemblyCode};
use rust_vm_lib::byte_code::ByteCodes;
use crate::disassembly_table::{DISASSEMBLY_TABLE, OPERATOR_DISASSEMBLY_TABLE, Argument};
use crate::error;


pub fn disassemble(byte_code: ByteCode, verbose: bool) -> AssemblyCode {

    let mut assembly = AssemblyCode::new();

    let mut i: usize = 0;
    while i < byte_code.len() {

        let start_index = i;
        let operator = byte_code[start_index];
        i += 1;

        let (base_name, sizes, args) 
            = DISASSEMBLY_TABLE.get(operator as usize)
            .unwrap_or_else(
                || error::invalid_instruction_code(operator, start_index)
        );   

        if verbose {
            print!("\n{}: {} ({})", start_index, ByteCodes::from(operator), operator);
        }
        
        let mut line = base_name.to_string();

        if let Some(args) = args {
            // The operator has arguments

            // Need mutable args to update them
            let mut args: Vec<Argument> = args.to_vec();

            if let Some(sizes) = sizes {
                // The operands have a variable size
                // The first byte is the handled size for sized operands
                let handled_size = byte_code[i];
                i += 1;
                line += &handled_size.to_string();

                // Update the arguments with variable sizes
                for size in sizes {
                    args[*size as usize].size = handled_size;
                }
            }

            // Disassemble the operands
            for arg in args {
                let disassembler = match OPERATOR_DISASSEMBLY_TABLE.get(arg.kind as usize) {
                    Some(disassembler) => disassembler,
                    None => panic!("Invalid argument type: {}. This is a bug.", arg.kind)
                };
                
                let operand_string = match disassembler(&byte_code[i..i + arg.size as usize]) {
                    Ok(operand_string) => operand_string,
                    Err(message) => error::invalid_arguments(&arg, &byte_code[i..i + arg.size as usize], i, &message)
                };
                
                i += arg.size as usize;

                line += &format!(" {}", operand_string);
            }

        }

        if verbose {
            println!(" <= {:?}\n{}\n", &byte_code[start_index..i], line);
        }

        assembly.push(line);
    }

    assembly
}

